apps/gdalwarp_lib.cpp

```diff
@@ -3873,22 +3873,72 @@ static GDALDatasetH GDALWarpCreateOutput(
                 }
             }
 
-            psInfo->pfnTransform(hTransformArg, TRUE, nPoints, &padfX[0],
-                                 &padfY[0], &padfZ[0], &pabSuccess[0]);
+            bool transformedToSrcCRS{false};
+
+            GDALGenImgProjTransformInfo *psTransformInfo{
+                static_cast<GDALGenImgProjTransformInfo *>(hTransformArg)};
+
+            // If a transformer is available, use an extent that covers the
+            // target extent instead of the real source image extent
+            if (psTransformInfo && psTransformInfo->pReprojectArg)
```

Comment:
We should also likely check that psTransformInfo->pSrcTransformer == nullptr to avoid the below logic to trigger when we use a image pixel -> source CRS coordinates transformer that is not geotransform-based (such as RPC, geolocation arrays, ...)

Reply:
done